examples/generative/gaugan.py

```diff
@@ -0,0 +1,792 @@
+    dataset = tf.data.Dataset.from_tensor_slices(
+        (image_files, segmentation_map_files, label_files)
+    )
+
+    dataset = dataset.shuffle(batch_size * 10) if is_train else dataset
+    dataset = dataset.map(_load_data_tf, num_parallel_calls=AUTOTUNE)
+    dataset = dataset.map(_random_crop, num_parallel_calls=AUTOTUNE)
+    dataset = dataset.map(
+        lambda x, y, z: (x, y, tf.one_hot(z, N_CLASS)), num_parallel_calls=AUTOTUNE
+    )
+    return dataset.batch(batch_size, drop_remainder=True)
+
+
+train_dataset = load(train_files, batch_size=BATCH_SIZE, is_train=True)
+val_dataset = load(val_files, batch_size=BATCH_SIZE, is_train=False)
+
+"""
+Let's now visualize a few samples from the training set. 
+"""
+
+sample_train_batch = next(iter(train_dataset))
+print(f"Segmentation map batch shape: {sample_train_batch[0].shape}.")
+print(f"Image batch shape: {sample_train_batch[1].shape}.")
+print(f"One-hot encoded label map shape: {sample_train_batch[2].shape}.")
+
+# Plot a view samples from the training set.
+for segmentation_map, real_image in zip(sample_train_batch[0], sample_train_batch[1]):
+    fig = plt.figure(figsize=(10, 10))
+    fig.add_subplot(1, 2, 1).set_title("Segmentation Map")
+    plt.imshow((segmentation_map + 1) / 2)
+    fig.add_subplot(1, 2, 2).set_title("Real Image")
+    plt.imshow((real_image + 1) / 2)
+    plt.show()
+
+"""
+Note that in the rest of this example, we have used a couple of figures from the
+[original GauGAN paper](https://arxiv.org/abs/1903.07291) for convenience. 
+"""
+
+"""
+# Custom layers
+
+In the following block we implement the following layers:
+
+* SPADE
+* Residual block including SPADE
+* Gaussian sampler
+"""
+
+"""
+### Some more notes on SPADE
+
+![](https://i.imgur.com/DgMWrrs.png)
+
+**SPatially-Adaptive (DE) normalization** or **SPADE** is a simple but effective layer
+for synthesizing photorealistic images given an input semantic layout. Previous methods
+for conditional image generation from semantic input such as
+Pix2Pix ([Isola et al.](https://arxiv.org/abs/1611.07004))
+or Pix2PixHD ([Wang et al.](https://arxiv.org/abs/1711.11585))
+directly feeds the semantic layout as input to the deep network, which is then processed
+through stacks of convolution, normalization, and nonlinearity layer. This is often
+suboptimal as the normalization layers have a tendency to wash away semantic information.
+
+In SPADE, the segmentation mask is first projected onto an embedding space and then
+convolved to produce the modulation parameters `γ` and `β`. Unlike prior conditional
+normalization methods, `γ` and `β` are not vectors, but tensors with spatial dimensions.
+The produced `γ` and `β` are multiplied and added to the normalized activation
+element-wise. As the modulation parameters are adaptive to the input segmentation mask,
+the proposed SPADE is better suited for semantic image synthesis.
+"""
+
+
+class SPADE(layers.Layer):
+    def __init__(self, filters, epsilon=1e-5):
```

Comment:
In general in custom layers, accept `**kwargs` in the signature and pass it to `super().__init__(**kwargs)`, which enables layer naming, etc.